docs: migrate 'interfaces' (#5464)
canonical/snapcraft: docs/explanation/interfaces.rst

Interfaces
==========

A strictly-confined snap is considered untrusted and runs in a restricted environment.
It's only able to access a limited set of resources outside the environment it runs in.
Access to system resources and other snaps is granted on a granular basis using a
mechanism called `interfaces <https://snapcraft.io/docs/interface-management>`_.

For example, a browser without network access doesn't serve its intended purpose.
To that end, snap developers can use the `network interface
<https://snapcraft.io/docs/network-interface>`_ to provide network access to the
browser.

Design
------

> .. image:: https://assets.ubuntu.com/v1/59c290a8-snapd-interfaces.png

Author: rant: I've never liked this image because:

* it doesn't show the same resource being provided to multiple consumers
* it only shows a snap providing a resource and not a system resource
* the direction of flow is "backwards" (it should flow from right to left or have an arrow)

Reviewer: Is it bad enough that we should exclude it completely for now?

Author: It's better than nothing because it does show the relationship between slots, plugs, interfaces so I guess we can keep it for now.  

Does it still not show up for you? I can see it.

Reviewer 2: All good now! Maybe it was something on my end.

   :alt: Interface between slots and plugs.

> An interface consists of a connection between a slot and a plug. A slot is
> the provider of the interface and may be a system resource or access to another snap.
> A slot can support multiple plug connections. A plug is the consumer and connects to a
> slot through an interface.

Reviewer 2: ```suggestion
A slot can be connected to one or more plugs through an interface. The slot provides
this interface and access to either a system resource or a resource from another snap.
The plug connects to the provided interface and consumes the resource.
```

Disregard if this is less accurate than the original.

Author: https://github.com/canonical/snapcraft/pull/5464/commits/aef48b8d2374797eea498351655725de6f646ff0

> Interfaces can be automatically or manually connected. Some interfaces automatically
> connect when a snap is installed. Others interfaces don't automatically connect,
> especially if they have access to sensitive resources like network control.
> See `Supported interfaces <https://snapcraft.io/docs/supported-interfaces>`_ for
> details on which interfaces can auto-connect.

Reviewer 2: ```suggestion
Some interfaces connect automatically when a snap is installed, while others, such as
those that have access to sensitive resources, need to be connected manually. See
`Supported interfaces <https://snapcraft.io/docs/supported-interfaces>`_ for details on
which interfaces connect automatically.
```

Author: https://github.com/canonical/snapcraft/pull/5464/commits/aef48b8d2374797eea498351655725de6f646ff0

Users have the option to manually control interfaces by connecting and disconnecting

> them using snapd. See `Interface management

Author: https://github.com/canonical/snapcraft/pull/5464/commits/aef48b8d2374797eea498351655725de6f646ff0
